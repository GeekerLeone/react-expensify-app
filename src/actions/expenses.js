import {v4 as uuid} from 'uuid'
import { database } from '../firebase/firebase';
import {ref, push, get, remove, update} from "firebase/database";

export const addExpense = (expense) =>({
    type: "ADD_EXPENSE",
    expense
    });

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return push(ref(database, `user/${uid}/expenses`), expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
};

export const removeExpense = ({ id } = {}) => ({
    type:  'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) =>{
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(database, `user/${uid}/expenses/${id}`)).then(()=>{
            dispatch(removeExpense({id}))
        })
    }

}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return update(ref(database, `user/${uid}/expenses/${id}`), updates).then(() =>{
            dispatch(editExpense(id, updates))
        })
    }
}

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpense = () =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return get(ref(database, `user/${uid}/expenses`)).then((snapshot)=>{
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
        
        
    }
}
