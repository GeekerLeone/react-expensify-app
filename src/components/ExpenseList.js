import React from "react";
import { connect } from "react-redux";
import ExpenseItem from "./ExpenseListItem";
import selectedExpense from '../selectors/expenses'
import ExpenseListFilter from "./ExpenseListFilter";

export const ExpenseList = (props) => (
    <div>
        <ExpenseListFilter />
        {props.expenses.length == 0? (
            <p>No expense</p>
        ):(
            props.expenses.map((expense) => (
            <ExpenseItem key = {expense.id} {...expense}/>
        ))
        )}
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectedExpense(state.expenses, state.filter) 
    };
}

export default connect(mapStateToProps)(ExpenseList);
