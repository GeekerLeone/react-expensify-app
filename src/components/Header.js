import React from "react";
import { Outlet, NavLink } from "react-router-dom";
const Header = () =>(
    <header>
        <h1>Expensify</h1>
        <nav>
        <NavLink to="/" className={({ isActive }) => isActive ? "is-active" : ''}>Dashboard</NavLink>
        <NavLink to='/create' className={({ isActive }) => isActive ? "is-active" : ''}>Create Expense</NavLink>
        <NavLink to='/help' className={({ isActive }) => isActive ? "is-active" : ''}>Help</NavLink>
        </nav>
        <Outlet />
    </header>
);
export default Header;