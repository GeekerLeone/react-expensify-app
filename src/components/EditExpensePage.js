import React from "react";
import { useParams } from "react-router-dom";
import { connect, useSelector } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {editExpense,  startRemoveExpense, startEditExpense} from '../actions/expenses';
import { useNavigate } from 'react-router-dom';



const EditExpensePage = (props) =>{
    const params = useParams()
    const navigate = useNavigate();
    const expense = useSelector((state) => state.expenses.find(expense => expense.id ===params.id));
    return(
    <div>
        <ExpenseForm 
        expense = {expense}
        onSubmit={(expense) => {
            props.dispatch(startEditExpense(params.id ,expense))
            navigate('/')
        }}
        />
        <button 
        onClick={() =>{
            props.dispatch(startRemoveExpense({id: params.id}))
            navigate('/')
        }}
        >Remove</button>
    </div>
    );
}



export default connect()(EditExpensePage);
