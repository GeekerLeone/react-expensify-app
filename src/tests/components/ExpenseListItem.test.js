import React from "react";
import { shallow } from "enzyme";
import expenses from '../fixtures/expenses'
import {expenseItem} from '../../components/ExpenseListItem'

test('should render ExpenseListItem correctly', () => {
    const wrapper = shallow(<expenseItem expenses={expenses[0]}/>)
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpenseListItem correctly with empty expense', () => {
    const wrapper = shallow(<expenseItem expenses={[]}/>)
    expect(wrapper).toMatchSnapshot();
})