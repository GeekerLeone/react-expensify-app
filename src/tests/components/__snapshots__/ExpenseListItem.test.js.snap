// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<expenseItem
  expenses={
    Object {
      "amount": 10,
      "createdAt": 0,
      "description": "Gum",
      "id": "0",
      "note": "",
    }
  }
/>
`;

exports[`should render ExpenseListItem correctly with empty expense 1`] = `
<expenseItem
  expenses={Array []}
/>
`;
